File: src/main/java/xyz/devvydont/smprpg/items/blueprints/fishing/FishBlueprint.java
package xyz.devvydont.smprpg.items.blueprints.fishing;

import io.papermc.paper.datacomponent.item.Consumable;
import io.papermc.paper.datacomponent.item.consumable.ConsumeEffect;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.jspecify.annotations.NonNull;
import xyz.devvydont.smprpg.items.CustomItemType;
import xyz.devvydont.smprpg.items.ItemClassification;
import xyz.devvydont.smprpg.items.base.CustomItemBlueprint;
import xyz.devvydont.smprpg.items.interfaces.IEdible;
import xyz.devvydont.smprpg.items.interfaces.IModelOverridden;
import xyz.devvydont.smprpg.items.interfaces.ISellable;
import xyz.devvydont.smprpg.services.ItemService;

import java.util.List;

import static xyz.devvydont.smprpg.util.formatting.ComponentUtils.create;

/**
 * Represents a blueprint for a basic fish that you can fish up.
 */
public class FishBlueprint extends CustomItemBlueprint implements IModelOverridden, ISellable, IEdible {

    public FishBlueprint(ItemService itemService, CustomItemType type) {
        super(itemService, type);
    }

    /**
     * Get the base worth of this fish before applying a rarity modifier on it.
     * This essentially maps the COMMON worth of this item.
     * @return The base worth.
     */
    public static int getBaseWorth(CustomItemType fish) {
        return switch (fish) {
            case COD -> 75;
            case CARP -> 55;
            case SALMON -> 100;
            case GUPPY -> 120;

            case SNAPPER -> 450;
            case CATFISH -> 600;
            case PUFFERFISH -> 500;
            case CLOWNFISH -> 800;
            case BASS -> 1_000;

            case YELLOWFIN_TUNA -> 7_500;
            case BARRACUDA -> 4_000;
            case PIKE -> 4_500;
            case STURGEON -> 5_500;
            case BLUE_TANG -> 7_000;

            case GOLIATH_GROUPER -> 35_000;
            case LEAFY_SEADRAGON -> 30_000;
            case LIONFISH -> 25_000;

            case BLUE_MARLIN -> 90_000;
            case FANGTOOTH -> 100_000;
            case DEEP_SEA_ANGLERFISH -> 125_000;

            case BLISTERFISH -> 180;
            case IMPLING -> 215;

            case CRIMSONFISH -> 750;

            case BONE_MAW -> 4500;
            case SOUL_SCALE -> 7500;

            case FLAREFIN -> 50000;
            case GHOST_FISH -> 60000;

            case DEVIL_RAY -> 200000;

            case VOIDFIN -> 225;
            case ORBLING -> 300;

            case WARPER -> 1500;
            case BLOBFISH -> 2000;

            case GOBLIN_SHARK -> 40000;

            case STARSURFER -> 175000;
            case ABYSSAL_SQUID -> 250000;

            case TWILIGHT_ANGLERFISH -> 500000;
            case COSMIC_CUTTLEFISH -> 750000;

            default -> 50;
        };
    }

    /**
     * Maps fish types to the amount of nutrition (hunger bars) that they replenish.
     * @param fish The fish type.
     * @return The amount of half hunger bars to replenish.
     */
    public static int getNutrition(CustomItemType fish) {
        // Lazy for now, just use the rarity.
        return (fish.DefaultRarity.ordinal() + 1) * 2;
    }

    /**
     * Maps fish types to the amount of saturation (healing efficiency and delayed hunger decay) that they replenish.
     * @param fish The fish type.
     * @return The amount of saturation.
     */
    public static int getSaturation(CustomItemType fish) {
        // Lazy for now, just use the rarity.
        return (fish.DefaultRarity.ordinal() + 1) * 2;
    }

    /**
     * Because we don't want fish to interact with furnaces, all custom fish should be clownfish items under the hood.
     * We can however overwrite what the fish looks like here. If we don't define a material, we are going to assume
     * we want the item to display as defined in {@link CustomItemType}.
     * @return The material texture override.
     */
    @Override
    public Material getDisplayMaterial() {
        return switch (this.getCustomItemType()) {
            // case CATFISH -> Material.CAT_SPAWN_EGG;
            // case LIONFISH -> Material.FOX_SPAWN_EGG;
            // case COD, PIKE, GUPPY -> Material.COD;
            // case SALMON, BLISTERFISH -> Material.SALMON;
            // case PUFFERFISH, VOIDFIN -> Material.PUFFERFISH;
            // case CLOWNFISH -> Material.TROPICAL_FISH;
            default -> this.getCustomItemType().DisplayMaterial;
        };
    }

    @Override
    public int getWorth(ItemStack item) {
        return getBaseWorth(this.getCustomItemType()) * item.getAmount();
    }

    @Override
    public ItemClassification getItemClassification() {
        return ItemClassification.CONSUMABLE;
    }

    @Override
    public boolean canAlwaysEat(ItemStack item) {
        return false;
    }

    @Override
    public int getNutrition(ItemStack item) {
        return getNutrition(this.getCustomItemType());
    }

    @Override
    public float getSaturation(ItemStack item) {
        return getSaturation(this.getCustomItemType());
    }

    @Override
    public @NonNull Consumable getConsumableComponent(ItemStack item) {
        var component = Consumable.consumable()
                .consumeSeconds(1.6f);

        // Conditionally add effects for any "weird" fish.
        if (this.getCustomItemType() == CustomItemType.PUFFERFISH)
            component.addEffect(ConsumeEffect.applyStatusEffects(
                    List.of(new PotionEffect(PotionEffectType.POISON, 20, 0, false, false)),
                    1.0f
            ));

        return component.build();
    }
}

File: src/main/java/xyz/devvydont/smprpg/items/blueprints/fishing/FishBlueprint.kt
package xyz.devvydont.smprpg.items.blueprints.fishing

import io.papermc.paper.datacomponent.item.Consumable
import io.papermc.paper.datacomponent.item.consumable.ConsumeEffect
import net.kyori.adventure.key.Key
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType
import xyz.devvydont.smprpg.items.CustomItemType
import xyz.devvydont.smprpg.items.ItemClassification
import xyz.devvydont.smprpg.items.base.CustomItemBlueprint
import xyz.devvydont.smprpg.items.interfaces.IEdible
import xyz.devvydont.smprpg.items.interfaces.IModelOverridden
import xyz.devvydont.smprpg.items.interfaces.ISellable
import xyz.devvydont.smprpg.services.ItemService

/**
 * Represents a blueprint for a basic fish that you can fish up.
 */
class FishBlueprint(itemService: ItemService, type: CustomItemType) : CustomItemBlueprint(itemService, type),
    IModelOverridden, ISellable, IEdible {

    override val itemClassification: ItemClassification get() = ItemClassification.CONSUMABLE

    /**
     * Because we don't want fish to interact with furnaces, all custom fish should be clownfish items under the hood.
     * We can however overwrite what the fish looks like here. If we don't define a material, we are going to assume
     * we want the item to display as defined in [CustomItemType].
     * @return The material texture override.
     */
    override fun getDisplayKey(): Key { return IModelOverridden.ofItemTypeInDirectory(customItemType, "consumables/fish") }

    override fun getWorth(item: ItemStack): Int { return getBaseWorth(this.customItemType) * item.amount }

    override fun canAlwaysEat(item: ItemStack?): Boolean { return false }

    override fun getNutrition(item: ItemStack?): Int { return getNutrition(this.customItemType) }

    override fun getSaturation(item: ItemStack?): Float { return getSaturation(this.customItemType).toFloat() }

    override fun getConsumableComponent(item: ItemStack?): Consumable {
        val component = Consumable.consumable()
            .consumeSeconds(1.6f)

        // Conditionally add effects for any "weird" fish.
        if (this.customItemType == CustomItemType.PUFFERFISH) component.addEffect(
            ConsumeEffect.applyStatusEffects(
                listOf(PotionEffect(PotionEffectType.POISON, 20, 0, false, false)),
                1.0f
            )
        )

        return component.build()
    }

    companion object {
        /**
         * Get the base worth of this fish before applying a rarity modifier on it.
         * This essentially maps the COMMON worth of this item.
         * @return The base worth.
         */
        fun getBaseWorth(fish: CustomItemType): Int {
            return when (fish) {
                CustomItemType.COD -> 75
                CustomItemType.CARP -> 55
                CustomItemType.SALMON -> 100
                CustomItemType.GUPPY -> 120
                CustomItemType.SNAPPER -> 450
                CustomItemType.CATFISH -> 600
                CustomItemType.PUFFERFISH -> 500
                CustomItemType.CLOWNFISH -> 800
                CustomItemType.BASS -> 1000
                CustomItemType.YELLOWFIN_TUNA -> 7500
                CustomItemType.BARRACUDA -> 4000
                CustomItemType.PIKE -> 4500
                CustomItemType.STURGEON -> 5500
                CustomItemType.BLUE_TANG -> 7000
                CustomItemType.GOLIATH_GROUPER -> 35000
                CustomItemType.LEAFY_SEADRAGON -> 30000
                CustomItemType.LIONFISH -> 25000
                CustomItemType.BLUE_MARLIN -> 90000
                CustomItemType.FANGTOOTH -> 100000
                CustomItemType.DEEP_SEA_ANGLERFISH -> 125000
                CustomItemType.BLISTERFISH -> 180
                CustomItemType.IMPLING -> 215
                CustomItemType.CRIMSONFISH -> 750
                CustomItemType.BONE_MAW -> 4500
                CustomItemType.SOUL_SCALE -> 7500
                CustomItemType.FLAREFIN -> 50000
                CustomItemType.GHOST_FISH -> 60000
                CustomItemType.DEVIL_RAY -> 200000
                CustomItemType.VOIDFIN -> 225
                CustomItemType.ORBLING -> 300
                CustomItemType.WARPER -> 1500
                CustomItemType.BLOBFISH -> 2000
                CustomItemType.GOBLIN_SHARK -> 40000
                CustomItemType.STARSURFER -> 175000
                CustomItemType.ABYSSAL_SQUID -> 250000
                CustomItemType.TWILIGHT_ANGLERFISH -> 500000
                CustomItemType.COSMIC_CUTTLEFISH -> 750000
                else -> 50
            }
        }

        /**
         * Maps fish types to the amount of nutrition (hunger bars) that they replenish.
         * @param fish The fish type.
         * @return The amount of half hunger bars to replenish.
         */
        fun getNutrition(fish: CustomItemType): Int {
            // Lazy for now, just use the rarity.
            return (fish.DefaultRarity.ordinal + 1) * 2
        }

        /**
         * Maps fish types to the amount of saturation (healing efficiency and delayed hunger decay) that they replenish.
         * @param fish The fish type.
         * @return The amount of saturation.
         */
        fun getSaturation(fish: CustomItemType): Int {
            // Lazy for now, just use the rarity.
            return (fish.DefaultRarity.ordinal + 1) * 2
        }
    }
}
